### PR TITLE
Fix PersistentStorageDelegate 0-storage behavior

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
#### Problem
PersistentStorageDelegate's backends do not deal with
nullptr input yet in SDK.  This can be addressed with pre-checks
in the default PersistentStorageDelegate implementations
in the short term.

Issue #16130

#### Change overview
- Make Server's PersistentStorageDelegate and
  TestPersistentStorage delegate properly handle nullptr
  input for zero-length.
- Document behavior expected of PersistentStorageDelegate interface

#### Testing

- Unit tests still pass
- Cert tests still pass

### CODE REVIEW

**Review by emargolis — APPROVED** [review]

File: src/lib/core/CHIPPersistentStorageDelegate.h [hunk 1]
```diff
@@ -62,11 +62,15 @@ class DLL_EXPORT PersistentStorageDelegate
 
     /**
      * @brief
-     *   Set the value for the key to a byte buffer.
+     *   Set the value for the key to a byte buffer. Empty keys can be stored
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPPersistentStorageDelegate.h:65]:
```suggestion
     *   Set the value for the key to a byte buffer. Empty values can be stored
```

File: src/lib/core/CHIPPersistentStorageDelegate.h [hunk 2]
```diff
@@ -62,11 +62,15 @@ class DLL_EXPORT PersistentStorageDelegate
...
+     *   with size == 0.
      *
      * @param[in] key Key to be set
-     * @param[in] value Value to be set
-     * @param[in] size Size of the Value
+     * @param[in] value Pointer to buytes of value to be set. `value` can only be `nullptr` if size == 0.
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPPersistentStorageDelegate.h:69]:
```suggestion
     * @param[in] value Pointer to bytes of value to be set. `value` can only be `nullptr` if size == 0.
```

File: src/lib/support/TestPersistentStorageDelegate.h [hunk 3]
```diff
@@ -45,12 +46,19 @@ class TestPersistentStorageDelegate : public PersistentStorageDelegate
...
+        size_t valueSize             = value.size();
+        if (size < valueSize)
+        {
+            size = CanCastTo<uint16_t>(valueSize) ? static_cast<uint16_t>(valueSize) : 0;
+            return CHIP_ERROR_BUFFER_TOO_SMALL;
+        }
+        else
+        {
+            size = static_cast<uint16_t>(valueSize);
+            VerifyOrReturnError(buffer != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
```
Comment (bzbarsky-apple) [inline src/lib/support/TestPersistentStorageDelegate.h:58]: This does not seem consistent with the Server.h delegate impl in terms of what happens if buffer is null and the incoming size is 0...  And the API documentation does not say what will happen in that case.

File: src/lib/support/tests/TestTestPersistentStorageDelegate.cpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    size = static_cast<uint16_t>(strlen(kStringValue2) - 1);
+    err = storage.SyncGetKeyValue("key2", &buf[0], size);
+    NL_TEST_ASSERT(inSuite, err == CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    // Make sure we get told of the right amount to read
+    NL_TEST_ASSERT(inSuite, size == strlen(kStringValue2));
+
+    // Read in too small a buffer, which is nullptr (i.e. just try to find if key exists)
+    size = static_cast<uint16_t>(strlen(kStringValue2) - 1);
+    err = storage.SyncGetKeyValue("key2", nullptr, size);
```
Comment (bzbarsky-apple) [inline src/lib/support/tests/TestTestPersistentStorageDelegate.cpp:112]: This looks like an invalid call to me (passing nullptr and a nonzero length)...

File: src/lib/support/TestPersistentStorageDelegate.h (hunk 3 above)
Comment (tcarmelveilleux) [reply src/lib/support/TestPersistentStorageDelegate.h:58]: Will fix the docs

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**